lexers/typst.lua

```diff
@@ -0,0 +1,125 @@
+    field = lex:tag(lexer.IDENTIFIER, hash_word) *
+            lex:tag(lexer.OPERATOR, P('.')) *
+            lex:tag('FIELD', lexer.word) * -S('[('),
+    operator = lex:tag(lexer.OPERATOR, S('+-/*%<>~!=^&|?~:;,.()[]{}')),
+    label = -B('\\') * lex:tag(lexer.LABEL, P('<') * lexer.word * P('>')),
+    label_two = -B('\\') * lex:tag(lexer.LABEL, P('@') * lexer.word),
+    link = P('http') * P('s')^-1 * P(':') * (lexer.word + S('.:/'))^1,
+
+    math = -B('\\') * lexer.range('$', false, false),
+    code = lexer.range('```', '```', false),
+    list = lex:tag(lexer.LIST, lexer.starts_line(lexer.digit^1 * '.' + S('+-'), true) * S(' \t')),
+    -- TODO: Do we really need to not tag a number if procceded by an alpha
+    comment = lex:tag(lexer.COMMENT, lexer.range('/*', '*/') + lexer.to_eol('//')),
+
+    keyword = lex:tag(lexer.KEYWORD, keyword_match),
+
+    header = header(6) + header(5) + header(4) + header(3) + header(2) + header(1)
+  }
+end
+
+local emb_lex = lexer.new('scripting')
+
+--[[
+ #{ ... }
+   OR
+ #let x = { ... }
+]]
+-- This is very limited, since it would only work correctly if no nested structures (w/ brackets) are found inside
+-- otherwise (if they're found inside), the first closing bracket of that nested structure would close the whole embedded
+-- script, causing the rest of the script to not be treated as a part of embedded script
+local start = (lex:tag(lexer.KEYWORD, P('#') * lex:word_match(lexer.KEYWORD)) *
+	      #((lexer.any - S('{;\n'))^1 * S('{') * lexer.space^0)) +
+	      lex:tag(lexer.OPERATOR,P('#') * S('{'))
+local embed_start = lex:tag('emb_tag', start)
+local embed_end = lexer:tag('emb_tag', S('}'))
+
+local function add_rules(lexer_obj, pre)
+  local rules = build_rules(pre)
+  lexer_obj:add_rule('header', rules.header)
+  lexer_obj:add_rule('field', rules.field)
+  lexer_obj:add_rule('function', rules.mod_func + rules.func)
+  lexer_obj:add_rule('method', rules.method)
+  lexer_obj:add_rule('label', rules.label + rules.label_two)
+  lexer_obj:add_rule('code', lex:tag(lexer.CODE, rules.code))
+  lexer_obj:add_rule('string', lex:tag(lexer.STRING, rules.string))
+  lexer_obj:add_rule('link', lex:tag(lexer.LINK, rules.link))
+  lexer_obj:add_rule('math', lex:tag('environment.math', rules.math))
+  lexer_obj:add_rule('keyword', rules.keyword)
+  lexer_obj:add_rule('identifier', rules.iden)
+  --lexer_obj:add_rule('number', lex:tag(lexer.NUMBER, rules.numeric_value))
+  lexer_obj:add_rule('list', rules.list)
+  lexer_obj:add_rule('comment', rules.comment)
+  lexer_obj:add_rule('operator', rules.operator)
+end
+
+-- Keywords, functions... don't need '#' when in code
+-- the character `#` is not valid in code
+-- TODO: only enable styling and text related rules when in []
+add_rules(emb_lex, '')
+
+lex:embed(emb_lex, embed_start, embed_end)
+
+add_rules(lex, '#')
+
+
+lex:set_word_list(lexer.KEYWORD, {
+  'if', 'else', 'for', 'while', 'let', 'set', 'import', 'include', 'return',
```

Comment:
It's a bit of a nitpick for now, but in my brief testing,

```
#while n < 10 {
  n = (n * 2) - 1
  (n,)
}
```

the `#while` was not highlighting as a keyword.

I just thought I'd make a note.